src/components/App/App.js

```diff
@@ -0,0 +1,288 @@
+import ProtectedRoute from "../../ProtectedRoute/ProtectedRoute";
+
+const App = () => {
+  const [userLoginModal, setUserLoginModal] = useState(false);
+  const [userRegisterModal, setUserRegisterModal] = useState(false);
+  const [isLoading, setIsLoading] = useState(false);
+  const [isSearching, setIsSearching] = useState(false);
+  const [currentUser, setCurrentUser] = useState(null);
+  const [newsCards, setNewsCards] = useState({});
+  const [visible, setVisible] = useState(3);
+  const [isSuccessfulModal, setIsSuccessfulModal] = useState(false);
+  const [isHomePage, setIsHomePage] = useState(false);
+  const [authError, setAuthError] = useState("");
+  const [savedArticles, setSavedArticles] = useState([]);
+  const [selectedArticleId, setSelectedArticleId] = useState(null);
+  const [keyword, setKeyword] = useState(null);
+
+  const history = useHistory();
+  const token = localStorage.getItem("jwt");
+
+  const handleVisibleReset = () => {
+    setVisible(3);
+  };
+
+  // BUTTON CLICK FUNCTIONS
+  const handleCloseModal = () => {
+    setUserRegisterModal(false);
+    setUserLoginModal(false);
+    setIsSuccessfulModal(false);
+  };
+
+  const handleLoginModal = () => {
+    setUserLoginModal(true);
+  };
+
+  const handleSignOutClick = () => {
+    setCurrentUser(null);
+    localStorage.clear();
+    setIsHomePage(true);
+    history.push("/");
+  };
+
+  const handleSavedNewsEnter = () => {
+    setIsHomePage(false);
+  };
+
+  const handleSavedNewsExit = () => {
+    setIsHomePage(true);
+    setNewsCards({});
+  };
+
+  const showMoreItems = () => {
+    setVisible((items) => items + 3);
+  };
+
+  // API FUNCTIONS
+
+  const handleSearchArticles = (input) => {
+    const keyword =
+      input.charAt(0).toUpperCase() + input.replace(/ .*/, "").slice(1);
+    setKeyword(keyword);
+    setIsLoading(true);
+    setIsSearching(true);
+    Api.search({ input })
+      .then((data) => {
+        setNewsCards(data.articles);
+        localStorage.setItem("articles", JSON.stringify(data.articles));
+        localStorage.setItem("keyword", keyword);
+      })
+      .catch((e) => {
+        console.error("Error while searching articles:", e);
+      })
+      .finally(() => {
+        setIsLoading(false);
+        setIsSearching(false);
+        setVisible(3);
+      });
+  };
+
+  const handleLogin = (inputs) => {
+    setIsLoading(true);
+    MainApi.signIn(inputs)
+      .then((data) => {
+        if (data.token) {
+          localStorage.setItem("jwt", data.token);
+          getSavedNews(data.token);
+          setUserLoginModal(false);
+        } else {
+          setAuthError(data.message || "Invalid data");
+        }
+      })
+      .catch((e) => {
+        if (String(e).includes("401") || String(e).includes("400")) {
+          setAuthError("Incorrect email or password");
+        } else {
+          setAuthError("An error occurred while signing in");
+        }
+      })
+      .finally(() => {
+        setIsLoading(false);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js
  function handleSubmit(request) {
    // start loading
    setIsLoading(true);
    request()
      // we need to close only in `then`
      .then(closeAllPopups)
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
      .catch(console.error)
      // and in finally we need to stop loading
      .finally(() => setIsLoading(false));
  }

```


You need to pass the **request** into the call of `handleSubmit` like that:

```js
 // here is an example
 function handleProfileFormSubmit(inputValues) {
  // here we create a function that returns a promise 
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally`
    return api.editProfile(inputValues).then(setCurrentUser);
  }
  // here we call handleSubmit passing the request
  handleSubmit(makeRequest);
}


```

So, this way you can remove a lot of code duplicating